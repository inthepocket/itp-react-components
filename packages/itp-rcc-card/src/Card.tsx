import * as React from 'react';
import classnames from 'classnames';
import CardContext from './CardContext';
import CardStylesInterface from './CardStylesInterface';
import CardBody from './CardBody';
import CardFooter from './CardFooter';
import CardHeader from './CardHeader';

export interface CardPropsInterface {
  children?: React.ReactNode | Element;
  href?: string;
  prefixCss?: string;
  size: 'small' | 'default' | 'large';
  styles?: CardStylesInterface;
  onClick?: ((event: React.MouseEvent<HTMLElement>) => void);
  target?: string;
}

interface CardComponentPropsInterface {
  className?: string;
  href?: string;
  onClick?: ((event: React.MouseEvent<HTMLElement>) => void);
  target?: string;
};

const { Provider } = CardContext;

/**
 * Card
 */
class Card extends React.PureComponent<CardPropsInterface> {
  static defaultProps = {
    prefixCss: 'card',
    target: '_self',
  };

  static Body = CardBody;
  static Header = CardHeader;
  static Footer = CardFooter;

  render() {
    const { children, href, prefixCss, onClick, size, styles, target } = this.props;
    const containerClassName = styles ? styles.card : prefixCss;
    const linkClassName = styles ? styles.link : `${prefixCss}-link`;

    const className = classnames({
      [`${containerClassName}`]: containerClassName,
      [`${linkClassName}`]: href && linkClassName,
    });

    const cardComponentProps: CardComponentPropsInterface = {
      className,
    };

    if (href) {
      cardComponentProps.href = href;

      if (target) {
        cardComponentProps.target = target;
      }
    }

    if (onClick) {
      cardComponentProps.onClick = onClick;
    }

    const cardComponent = href ? <a {...cardComponentProps}>{children}</a> :
      <div {...cardComponentProps}>{children}</div>;

    return (
      <Provider value={{ prefixCss, size, styles }}>
        {cardComponent}
      </Provider>
    );
  }
}

export default Card;
