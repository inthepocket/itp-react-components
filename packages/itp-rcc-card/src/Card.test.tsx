import * as React from 'react';
import { create } from 'react-test-renderer';
import Card from './Card';

const cssModuleMock = {
  body: 'cssModulesBody',
  bodySizeLarge: 'cssModulesBodySizeLarge',
  bodySizeSmall: 'cssModulesBodySizeSmall',
  card: 'cssModulesCard',
  footer: 'cssModulesFooter',
  footerSizeLarge: 'cssModulesFooterSizeLarge',
  footerSizeSmall: 'cssModulesFooterSizeSmall',
  header: 'cssModulesHeader',
  headerSizeLarge: 'cssModulesHeaderSizeLarge',
  headerSizeSmall: 'cssModulesHeaderSizeSmall',
  typePrimary: 'cssModulesTypePrimary',
  typeSecondary: 'cssModulesTypeSecondary',
  typeTertiary: 'cssModulesTypeTertiary',
};

const CardBody = () => (<p>Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.</p>);
const CardFooter = () => (<p>Quis nostrum exercitationem ullam corporis suscipit.</p>);
const CardHeader = () => (<h1>Excepteur sint occaecat cupidatat</h1>);
const CardBodyWithTitle = () => (
  <React.Fragment>
    <h2>Ullamco laboris nisi ut</h2>
    <CardBody />
  </React.Fragment>
);

test('Card: default', () => {
  const tree = create((
    <Card size="default">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: linked', () => {
  const tree = create((
    <Card href="https://inthepocket.mobi/" size="default">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: linked with target', () => {
  const tree = create((
    <Card href="https://inthepocket.mobi/" size="default" target="_blank">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: onClick', () => {
  const tree = create((
    <Card onClick={() => 'foo:bar'} size="default">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree.props.onClick()).toBe('foo:bar');
  expect(tree).toMatchSnapshot();
});

test('Card: size: small', () => {
  const tree = create((
    <Card size="small">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: size: large', () => {
  const tree = create((
    <Card size="large">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: type: primary', () => {
  const tree = create((
    <Card type="primary">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: type: secondary', () => {
  const tree = create((
    <Card type="secondary">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: type: tertiary', () => {
  const tree = create((
    <Card type="tertiary">
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: css modules: default', () => {
  const tree = create((
    <Card size="default" styles={cssModuleMock}>
      <Card.Header>
        <CardHeader />
      </Card.Header>
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
      <Card.Footer>
        <CardFooter />
      </Card.Footer>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: css modules: size: small', () => {
  const tree = create((
    <Card size="small" styles={cssModuleMock}>
      <Card.Header>
        <CardHeader />
      </Card.Header>
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
      <Card.Footer>
        <CardFooter />
      </Card.Footer>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: css modules: size: large', () => {
  const tree = create((
    <Card size="large" styles={cssModuleMock}>
      <Card.Header>
        <CardHeader />
      </Card.Header>
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
      <Card.Footer>
        <CardFooter />
      </Card.Footer>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: css modules: type: primary', () => {
  const tree = create((
    <Card type="primary" styles={cssModuleMock}>
      <Card.Header>
        <CardHeader />
      </Card.Header>
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
      <Card.Footer>
        <CardFooter />
      </Card.Footer>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: css modules: type: secondary', () => {
  const tree = create((
    <Card type="secondary" styles={cssModuleMock}>
      <Card.Header>
        <CardHeader />
      </Card.Header>
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
      <Card.Footer>
        <CardFooter />
      </Card.Footer>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: css modules: type: tertiary', () => {
  const tree = create((
    <Card type="tertiary" styles={cssModuleMock}>
      <Card.Header>
        <CardHeader />
      </Card.Header>
      <Card.Body>
        <CardBodyWithTitle />
      </Card.Body>
      <Card.Footer>
        <CardFooter />
      </Card.Footer>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: with header', () => {
  const tree = create((
    <Card size="default">
      <Card.Header>
        <CardHeader />
      </Card.Header>
      <Card.Body>
        <CardBody />
      </Card.Body>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: with footer', () => {
  const tree = create((
    <Card size="default">
      <Card.Body>
        <CardBody />
      </Card.Body>
      <Card.Footer>
        <CardFooter />
      </Card.Footer>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Card: with header and footer', () => {
  const tree = create((
    <Card size="default">
      <Card.Header>
        <CardHeader />
      </Card.Header>
      <Card.Body>
        <CardBody />
      </Card.Body>
      <Card.Footer>
        <CardFooter />
      </Card.Footer>
    </Card>
  ))
  .toJSON();

  expect(tree).toMatchSnapshot();
});
