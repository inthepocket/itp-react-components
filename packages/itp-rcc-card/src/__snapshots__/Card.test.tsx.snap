// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card: css modules: default 1`] = `
<div
  className="cssModulesCard"
>
  <header
    className="cssModulesHeader"
  >
    <h1>
      Excepteur sint occaecat cupidatat
    </h1>
  </header>
  <div
    className="cssModulesBody"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
  <footer
    className="cssModulesFooter"
  >
    <p>
      Quis nostrum exercitationem ullam corporis suscipit.
    </p>
  </footer>
</div>
`;

exports[`Card: css modules: size: large 1`] = `
<div
  className="cssModulesCard"
>
  <header
    className="cssModulesHeader cssModulesHeaderSizeLarge"
  >
    <h1>
      Excepteur sint occaecat cupidatat
    </h1>
  </header>
  <div
    className="cssModulesBody cssModulesBodySizeLarge"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
  <footer
    className="cssModulesFooter cssModulesFooterSizeLarge"
  >
    <p>
      Quis nostrum exercitationem ullam corporis suscipit.
    </p>
  </footer>
</div>
`;

exports[`Card: css modules: size: small 1`] = `
<div
  className="cssModulesCard"
>
  <header
    className="cssModulesHeader cssModulesHeaderSizeSmall"
  >
    <h1>
      Excepteur sint occaecat cupidatat
    </h1>
  </header>
  <div
    className="cssModulesBody cssModulesBodySizeSmall"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
  <footer
    className="cssModulesFooter cssModulesFooterSizeSmall"
  >
    <p>
      Quis nostrum exercitationem ullam corporis suscipit.
    </p>
  </footer>
</div>
`;

exports[`Card: default 1`] = `
<div
  className="card"
>
  <div
    className="card-body card-body-size-default"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
</div>
`;

exports[`Card: linked 1`] = `
<a
  className="card card-link"
  href="https://inthepocket.mobi/"
  target="_self"
>
  <div
    className="card-body card-body-size-default"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
</a>
`;

exports[`Card: linked with target 1`] = `
<a
  className="card card-link"
  href="https://inthepocket.mobi/"
  target="_blank"
>
  <div
    className="card-body card-body-size-default"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
</a>
`;

exports[`Card: onClick 1`] = `
<div
  className="card"
  onClick={[Function]}
>
  <div
    className="card-body card-body-size-default"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
</div>
`;

exports[`Card: size: large 1`] = `
<div
  className="card"
>
  <div
    className="card-body card-body-size-large"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
</div>
`;

exports[`Card: size: small 1`] = `
<div
  className="card"
>
  <div
    className="card-body card-body-size-small"
  >
    <h2>
      Ullamco laboris nisi ut
    </h2>
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
</div>
`;

exports[`Card: with footer 1`] = `
<div
  className="card"
>
  <div
    className="card-body card-body-size-default"
  >
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
  <footer
    className="card-footer card-footer-size-default"
  >
    <p>
      Quis nostrum exercitationem ullam corporis suscipit.
    </p>
  </footer>
</div>
`;

exports[`Card: with header 1`] = `
<div
  className="card"
>
  <header
    className="card-header card-header-size-default"
  >
    <h1>
      Excepteur sint occaecat cupidatat
    </h1>
  </header>
  <div
    className="card-body card-body-size-default"
  >
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
</div>
`;

exports[`Card: with header and footer 1`] = `
<div
  className="card"
>
  <header
    className="card-header card-header-size-default"
  >
    <h1>
      Excepteur sint occaecat cupidatat
    </h1>
  </header>
  <div
    className="card-body card-body-size-default"
  >
    <p>
      Cras sed ante. Phasellus in massa. Curabitur dolor eros, gravida et, hendrerit ac, cursus non, massa. Aliquam lorem.
    </p>
  </div>
  <footer
    className="card-footer card-footer-size-default"
  >
    <p>
      Quis nostrum exercitationem ullam corporis suscipit.
    </p>
  </footer>
</div>
`;
